### PR TITLE
[2.11.x] DDF-3473 The MetadataConfigurationParser does not respect the validUntil or cacheDuration attributes that might be present on SAML metadata

### PR DESCRIPTION
**Pull request by ahoffer**
#### What does this PR do?
1. Adds a cache duration to the IDP metadata XML file (duration = 1 day)
2. Adds a check, `MetadataConfigurationParser.validateMetadata` to ensure metadata has either a cache duration or a (future) valid-until date. 
3. More unit tests

**This is the first of what is expected to be 2 PRs. The next PR will update the consumers of the metadata to retrieve cache metadata using the duration/validUntil properties.**
#### Who is reviewing it? 
@vinamartin @mcalcote @rzwiefel @garrettfreibott 
#### Select relevant component teams: 
Security
https://github.com/orgs/codice/teams/security
#### Choose 2 committers to review/merge the PR. 

@clockard
@rzwiefel
@stustison

#### How should this be tested? (List steps with links to updated documentation)

Build with itests. 
Unzip distro and install DDF. 
Make sure login/logout works.

#### Any background context you want to provide?
Address tech debt and improves SAML compliance.

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-3473

#### Checklist:
- [NA] Documentation Updated
- [X] Update / Add Unit Tests
- [NA] Update / Add Integration Tests

#### Notes on Review Process
For the general case, merging is gated on the following:

* Review by at least two contributors
* Review by at least two committers
* Successful CI
* Successful static analysis (SonarQube/Coverity/etc.)
* Heroing of change (details vary by ticket)

Certain exceptions may occur. For example, if there is a problem with the CI environment then a manually performed full build and static analysis check may be substituted _at the discretion of the repository maintainers_. Additionally, repository metadata such as this `PULL_REQUEST_TEMPLATE`, the repository `README`, and other non-functional items may be merged without requiring CI and static analysis.

Finally, an abbreviated review process may be in place for forward-port tickets immediately following branching and prior to branch divergence. In that case, approvals, CI, static analysis, and heroing that occur against the PR against the _release branch_ can be considered sufficient for reviewing the PR against the _master branch_.

Once the branches diverge, the process will return to normal.

Insofar as divergence is not a function of time but of change to the parent branch, the repository maintainers will determine when abbreviated review ends.

When abbreviated review is in effect, a note to that effect will be at the top of this template. Following is an example note:
```
##### ABBREVIATED REVIEW BETWEEN 2.12.X AND MASTER IS IN EFFECT
____

```

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by ahoffer** [comment]
First build for this PR.
build now

**Comment by ahoffer** [comment]
The CI job crapped out, but it was successful:

 https://jenkins.codice.org/job/DDF-Jobs/job/2.11.x/job/pr/job/Linux/407/console

> ------------------------------------------------------------------------
>  BUILD SUCCESS
>  ------------------------------------------------------------------------
> [INFO] Total time: 45:49 min (Wall Clock)
> [INFO] Finished at: 2017-11-28T17:39:12+00:00
> INFO] Final Memory: 634M/2428M
> [INFO] ------------------------------------------------------------------------

**Comment by ahoffer** [comment]
Reran the test locally (Mac) with ` mvn verify it.test=TestApplicationService` and it passed. This may be a platform-specific error, or it might be a transient error. Re-running the CI build to attempt to reproduce the issue.
build now

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java [hunk 1]
```diff
@@ -69,6 +70,7 @@
   private static final String HTTP = "http://";
 
   private static final String FILE = "file:";
+  public static final Long ONE_DAY_MILLISECONDS = (long) 1000 * 3600 * 24;
```
Comment (clockard) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: Could also to `TimeUnit.HOURS.toMillis(24)`

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java [hunk 2]
```diff
@@ -238,7 +240,20 @@ private EntityDescriptor readEntityDescriptor(Reader reader) {
...
-    return (EntityDescriptor) entityXmlObj;
+  private void validateMetadata(EntityDescriptor root) {
+    boolean valid =
+        (root.getCacheDuration() != null && root.getCacheDuration() > 0)
+            || DateTime.now().isBefore(root.getValidUntil());
+    if (!valid) {
+      LOGGER.info(
+          "IDP metadata must either have cache duration or future valid-until date."
+              + " Setting cache duration to 24 hours.");
+      root.setCacheDuration(ONE_DAY_MILLISECONDS);
```
Comment (clockard) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: I am not very familiar with this part of the code but is it ok to just set the catch duration like this? Should we throw an exception instead?

**Comment by ahoffer** [comment]
Pushed new commit. Build now.

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java [hunk 3]
```diff
@@ -238,7 +240,19 @@ private EntityDescriptor readEntityDescriptor(Reader reader) {
...
     }
+    EntityDescriptor root = (EntityDescriptor) entityXmlObj;
+    validateMetadata(root);
+    return root;
+  }
 
-    return (EntityDescriptor) entityXmlObj;
+  private void validateMetadata(EntityDescriptor root) {
+    boolean cacheDurationPresent = root.getCacheDuration() != null && root.getCacheDuration() >= 0;
+    boolean datetimeStillValud = DateTime.now().isBefore(root.getValidUntil());
```
Comment (garrettfreibott) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: ✏️  Typo in variable "Valud" -> "Valid". Also maybe "datetime" -> "dateTime".

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java [hunk 4]
```diff
@@ -161,9 +178,45 @@ private EntityDescriptor getEntityDescriptor() {
...
+   */
+  public boolean isMetadataExpired() {
+    boolean missingCacheDuration = cacheDuration == null;
+    boolean missingValidUntil = validUntil == null;
+    boolean missingAllTimeConstraints = missingCacheDuration && missingValidUntil;
+    Instant now = Instant.now();
+    boolean cacheDurationGood = !missingCacheDuration && now.isBefore(created.plus(cacheDuration));
+    boolean stillValidUntil = !missingValidUntil && now.isBefore(validUntil);
+    boolean isMetadataGood = missingAllTimeConstraints || cacheDurationGood || stillValidUntil;
+    return !isMetadataGood;
```
Comment (rzwiefel) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java]: I'm not sure the logic in here is necessarily correct. According to the spec:
>Note that cache expiration does not imply a lack of validity in the absence of a `validUntil` attribute or 
other information; failure to update a cached instance (e.g., due to network failure) need not render 
metadata invalid, although implementations may offer such controls to deployers.
[Section 4.3.1](https://www.oasis-open.org/committees/download.php/56786/sstc-saml-metadata-errata-2.0-wd-05-diff.pdf)

>Metadata MUST be considered invalid upon reaching the time specified in a `validUntil` attribute of the 
subject element(s). The effective expiration may be adjusted downward by parent element(s) with earlier 
expirations. Invalid metadata MUST NOT be used. This contrasts with "stale" metadata that may be beyond its optimum cache duration but is not explicitly invalid. Such metadata remains valid and MAY be 
used at the discretion of the implementation.
[Section 4.3.2](https://www.oasis-open.org/committees/download.php/56786/sstc-saml-metadata-errata-2.0-wd-05-diff.pdf)

I think this is stating if it valid until is there and it is invalid, we MUST expire the metadata regardless of if caching is good. I would say we should also add defensive logic if both are set to ensure that the `validUntil` is respected even if both happen to be set. 

And according to the first sentence in section 4.3.1 
>Document caching MUST be based on the duration indicated by the `cacheDuration` attribute

This implies that we **cannot** cache if the cacheDuration is not set and must  fetch new metadata every time. 

We could try to get @stustison or maybe @coyotesqrl to double check as well if we wanted.

Comment (ahoffer) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java]: Looks like I was reading an outdated version of the spec. Yikes. :-(

**Comment by ahoffer** [comment]
build now

**Comment by ahoffer** [comment]
Fixed test. Build now.

**Comment by ahoffer** [comment]
Yarn failure. Rebuilding. Build now.

**Comment by ahoffer** [comment]
rebased again. build now

**Comment by ahoffer** [comment]
build now

**Comment by ahoffer** [comment]
Updated branch. Build now.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java [hunk 5]
```diff
@@ -61,7 +69,13 @@
 
   public void setMetadata(String metadata) {
     this.metadata = metadata;
+    eraseEntityDescriptions();
+  }
+
+  private void eraseEntityDescriptions() {
     entryDescriptions.getAndSet(null);
+    cacheDuration = null;
+    validUntil = null;
```
Comment (ahoffer) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java]: Need to null out created inst var

**Comment by ahoffer** [comment]
build now. 
[ERROR] Unable to download the NVD CVE data.[ERROR] No documents exist[ERROR] -> [Help 1][ERROR] [ERROR]

**Comment by ahoffer** [comment]
Had to rebase. Again. Build now.

**Comment by ahoffer** [comment]
New push. build now.

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java [hunk 6]
```diff
@@ -238,7 +238,18 @@ private EntityDescriptor readEntityDescriptor(Reader reader) {
...
+    validateMetadata(root);
+    return root;
+  }
 
-    return (EntityDescriptor) entityXmlObj;
+  private void validateMetadata(EntityDescriptor root) {
+    if (root.getCacheDuration() == null && root.getValidUntil() == null) {
+      LOGGER.info(
+          "IDP metadata must either have cache duration or valid-until date."
+              + " Defaulting IDP metadata cache duration to %s",
```
Comment (rzwiefel) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: this `%s` should be a `{}`

Comment (ahoffer) [reply platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: good catch

**Comment by ahoffer** [comment]
build now

**Comment by ahoffer** [comment]
The test failed because: 

> 22:04:06 Caused by: java.lang.AssertionError: Failed to start boot feature: Could not find the application named catalog-app. Start application failed

Because the changes were to the security app and not the catalog app I am going to try building again. If I get the same failure, I will debug it.

Build now

**Comment by LinkMJB** [comment]
`git fetch` retries caused GIB to lose context. Newer version of GIB is supposed to fix that problem.

build now

**Pull request merged**